### ISSUE TITLE
Memory consumption explodes using Bam I/O with millions of reference sequences 

### ISSUE THREAD

**Issue by rrahn** [seqan/seqan#1653]
It seems that our bam I/O consumes a lot of memory when dealing with Bam files containing large headers with millions of reference sequences.

**Comment by h-2** [comment]
@rrahn can you provide more feedback?

**Comment by temehi** [comment]
@rrahn "millions of reference sequences?" Is this an actual usage scenario or ?

**Comment by h-2** [comment]
@temehi yes, e.g. with protein databases.

**Comment by rrahn** [comment]
This was something @bkehr mentioned at the UGM 2016.
Well I have no specific test data set so we need to simulate it.

**Comment by bkehr** [comment]
We are mapping reads to a collection of several millions of contigs. When reading the headers of the resulting bam files with SeqAn, the memory consumption is very high. samtools view/sort needs much less memory.

As a side node: We mainly observed this with SeqAn 1.4, I am currently changing our code to use SeqAn 2.1.

**Comment by rrahn** [comment]
Ok maybe, we await you benchmarks result on 2.1 if they are still persistent. 
Thanks for the update.

**Comment by bkehr** [comment]
I tested this now on a bam file with 7771273 sequences listed in the header:

samtools view uses around 1 Gb
Seqan 1.4 uses around 30 Gb
SeqAn 2.1 uses around 23 Gb

**Comment by h-2** [comment]
We keep the full headers of all sequences in RAM.... can you tell us what the headers look like? Usually they can be truncated at first `' '` which saves **a lot** and most tools do this...

**Comment by bkehr** [comment]
Here are a few example lines from the header.

@SQ     SN:00.NODE_24228_length_150_cov_7.146667        LN:196
@SQ     SN:COMPONENT_0.2301_a_length_180_size_6 LN:180
@SQ     SN:COMPONENT_0.2302_a_length_293_size_11        LN:293
@SQ     SN:COMPONENT_0.2303_a_length_251_size_20        LN:251
@SQ     SN:COMPONENT_0.2304_a_length_337_size_18        LN:337
@SQ     SN:COMPONENT_0.2305_a_length_212_size_10        LN:212
@SQ     SN:COMPONENT_0.2306_a_length_231_size_5 LN:231
@SQ     SN:COMPONENT_0.2307_a_length_182_size_3 LN:182
@SQ     SN:COMPONENT_0.2308_a_length_123_size_10        LN:123
@SQ     SN:COMPONENT_0.2309_a_length_257_size_16        LN:257
@SQ     SN:COMPONENT_0.2310_a_length_176_size_27        LN:176
@SQ     SN:COMPONENT_0.2310_b_length_361_size_27        LN:361
@SQ     SN:COMPONENT_0.2311_a_length_209_size_10        LN:209

**Comment by rrahn** [comment]
@bkehr Can you verify, that only the header consumes this much memory?
Usually there is the seqNames and seqIds stored in the header plus some more additional lines. But I guess there aren't to many of them, right?
So 23 Gb seems to be a little off, wether truncated or not.

**Comment by h-2** [comment]
@rrahn No, it must be something else. If the headers are at most 80 characters wide and there are 8million of them that amounts to 610MB of memory...

**Comment by rrahn** [comment]
That's exactly, what I am saying.

**Comment by bkehr** [comment]
I made a test file `bam_header_test.bam` with 4 read alignments and the same header as above. The file itself is 141 MB. The following test program needs 23 GB of memory:

```
#include <seqan/bam_io.h>
using namespace seqan;

int main()
{
    BamFileIn infile("bam_header_test.bam");
    
    BamHeader header;
    readHeader(header, infile);

    return 0;
}
```

**Comment by h-2** [comment]
@bkehr can you provide us with the file? Either upload it to fu's ftp or just drop it on any of the local directories (/srv) on any of our servers, if you still have access! Thanks!

**Comment by bkehr** [comment]
Sorry I can't, but here is a simple way for you to generate a similar file yourself:

```
echo -e "@HD\tVN:1.4\tGO:none\tSO:coordinate" > test.sam
for i in {0000001..8000000}; do echo -e "@SQ\tSN:sequence_$i\tLN:500"; done >> test.sam
echo -e "READ1\t81\tsequence_0000007\t46\t60\t20M\tsequence_0000008\t28\t0\tATGCATGCATGCATGCATGC\tIIIIIIIIIIIIIIIIIIII" >> test.sam
echo -e "READ1\t161\tsequence_0000008\t28\t60\t20M\tsequence_0000007\t46\t0\tATGCATGCATGCATGCATGC\tIIIIIIIIIIIIIIIIIIII" >> test.sam
samtools view -Sb test.sam > bam_header_test.bam
```

**Comment by h-2** [comment]
fixed via #1752
